File: Exercise_4.java
// Time Complexity :O(N)
// Space Complexity: O(N)

class MergeSort 
{ 
    // Merges two subarrays of arr[]. 
    // First subarray is arr[l..m] 
    // Second subarray is arr[m+1..r] 
    void merge(int arr[], int l, int m, int r) 
    {  
        int leftsize = m -l +1;
        int rightsize = r-m;

        int[] leftarray = new int[leftsize];
        int[] rightarray = new int[rightsize];

       //Your code here  
       for(int i=0 ; i<leftsize ; i++)
       {
            leftarray[i] = arr[l+i];
       }

       for(int j=0 ; j<rightsize; j++)
       {
            rightarray[j] = arr[m+1+j];
       }

       int i=0, j=0, k=l;

       while(i < leftsize && j < rightsize)
       {
            if(leftarray[i] <= rightarray[j])
            {
                arr[k] = leftarray[i];
                i++;
            }
            else
            {
                arr[k] = rightarray[j];
                j++;
            }
            k++;
        }

        // loop thru and add the remaining elements in the left array to output array.
        while( i < leftsize)
        {
            arr[k] = leftarray[i];
            i++;
            k++;
        }

        //loop thru and add the remaining elements in the right array to the output array.
        while( j < rightsize)
        {
            arr[k] = rightarray[j];
            j++;
            k++;
        }
    } 
  
    // Main function that sorts arr[l..r] using 
    // merge() 
    void sort(int arr[], int l, int r) 
    { 
	    //Write your code here
        //Call mergeSort from here 

        if(l < r)
        {
            int mid = l+(r-l)/2;

            sort(arr, l , mid);
            sort(arr, mid+1, r);

            merge(arr, l, mid, r);
        }   
    } 
  
    /* A utility function to print array of size n */
    static void printArray(int arr[]) 
    { 
        int n = arr.length; 
        for (int i=0; i<n; ++i) 
            System.out.print(arr[i] + " "); 
        System.out.println(); 
    } 
  
    // Driver method 
    public static void main(String args[]) 
    { 
        int arr[] = {12, 11, 13, 5, 6, 7}; 
  
        System.out.println("Given Array"); 
        printArray(arr); 
  
        MergeSort ob = new MergeSort(); 
        ob.sort(arr, 0, arr.length-1); 
  
        System.out.println("\nSorted array"); 
        printArray(arr); 
    } 
} 

File: Exercise_4.cpp
// Time Complexity : O(n * log n)           
// Space Complexity : O(1)

#include<stdlib.h> 
#include<stdio.h> 
  
// Merges two subarrays of arr[]. 
// First subarray is arr[l..m] 
// Second subarray is arr[m+1..r] 
void merge(int arr[], int l, int m, int r) 
{ 
    int leftArraySize = m-l+1;
    int rightArraySize = r-m;
    //Your code here
    int leftArray[leftArraySize];
    int rightArray[rightArraySize];

    for(int i=0; i<leftArraySize; i++){
        leftArray[i] = arr[l+i];
    }
    for(int i=0; i<rightArraySize; i++){
        rightArray[i] = arr[m+1+i];
    }

    int i=0;
    int j=0;
    int k=l;

    while(i<leftArraySize && j<rightArraySize){
        if(leftArray[i] < rightArray[j]){
            arr[k] = leftArray[i];
            i++;
        } else {
            arr[k] = rightArray[j];
            j++;
        }
        k++;
    }

    while(i<leftArraySize){
        arr[k] = leftArray[i];
        i++;
        k++;
    }

    while(j<rightArraySize){
        arr[k] = rightArray[j];
        j++;
        k++;
    }
} 
  
/* l is for left index and r is right index of the 
   sub-array of arr to be sorted */
void mergeSort(int arr[], int left, int right) 
{ 
    if(left>=right){ return;}
    int m = (left+right)/2;
    
    mergeSort(arr, left, m);
    mergeSort(arr, m+1, right);
    
    merge(arr, left, m, right);

} 
  
/* UTILITY FUNCTIONS */
/* Function to print an array */
void printArray(int A[], int size) 
{ 
    int i; 
    for (i=0; i < size; i++) 
        printf("%d ", A[i]); 
    printf("\n"); 
} 
  
/* Driver program to test above functions */
int main() 
{ 
    //int arr[] = {12, 11, 13, 5, 6, 7}; 
    int arr[] = {12, 11,9, 8, 6, 5, 2, 1, 0}; 
    int arr_size = sizeof(arr)/sizeof(arr[0]); 
  
    printf("Given array is \n"); 
    printArray(arr, arr_size); 
  
    mergeSort(arr, 0, arr_size - 1); 
  
    printf("\nSorted array is \n"); 
    printArray(arr, arr_size); 
    return 0; 
}